pkg/cli/api_server.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cli
+
+import (
+	"context"
+	"fmt"
+
+	kms "cloud.google.com/go/kms/apiv1"
+	jvspb "github.com/abcxyz/jvs/apis/v0"
+	"github.com/abcxyz/jvs/internal/version"
+	"github.com/abcxyz/jvs/pkg/config"
+	"github.com/abcxyz/jvs/pkg/justification"
+	"github.com/abcxyz/pkg/cfgloader"
+	"github.com/abcxyz/pkg/cli"
+	"github.com/abcxyz/pkg/logging"
+	"github.com/abcxyz/pkg/serving"
+	"github.com/sethvargo/go-envconfig"
+	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/health"
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
+	"google.golang.org/grpc/reflection"
+)
+
+var _ cli.Command = (*APIServerCommand)(nil)
+
+type APIServerCommand struct {
+	cli.BaseCommand
+
+	// testLookuper overrides the lookuper. It is only used for testing.
+	testLookuper envconfig.Lookuper
+
+	// testBoundAddress is populated by the listener. It is only used for testing
+	// where the server is configured to listen on a raondom port, but tests need
+	// to know the port and address it chose.
+	testBoundAddress string
```

Comment:
These new pkg/cli files are almost identical to the `realMain` bodies from `cmd` with this exception:

- There's an ability to inject custom environments into the command during parsing (which propagates to envconfig).
- There's a private field which exposes the bind address. When services bind to port "0", the operating system selects a random, unused port. We use this to our advantage for parallel testing, where we can fully parallelize all tests by allowing the OS to pick a port, but we need to push the port back up the stack in the test so that clients can connect.